board/xen/xenguest_arm64/xenguest_arm64.c

```diff
@@ -0,0 +1,158 @@
+/*
+ * SPDX-License-Identifier: GPL-2.0+
+ *
+ * (C) 2013
+ * David Feng <fenghua@phytium.com.cn>
+ * Sharma Bhupesh <bhupesh.sharma@freescale.com>
+ *
+ * (C) 2020 EPAM Systems Inc
+ */
+
+#include <common.h>
+#include <cpu_func.h>
+#include <dm.h>
+#include <errno.h>
+#include <malloc.h>
+
+#include <asm/io.h>
+#include <asm/armv8/mmu.h>
+#include <asm/xen.h>
+#include <asm/xen/hypercall.h>
+
+#include <linux/compiler.h>
+
+DECLARE_GLOBAL_DATA_PTR;
+
+int board_init(void)
+{
+	return 0;
+}
+
+/*
+* Use fdt provided by Xen: according to
+* https://www.kernel.org/doc/Documentation/arm64/booting.txt
+* x0 is the physical address of the device tree blob (dtb) in system RAM.
+* This is stored in rom_pointer during low level init.
+*/
+void *board_fdt_blob_setup(void)
+{
+	if (fdt_magic(rom_pointer[0]) != FDT_MAGIC)
+		return NULL;
+	return (void *)rom_pointer[0];
+}
+
+#define MAX_MEM_MAP_REGIONS 5
+static struct mm_region xen_mem_map[MAX_MEM_MAP_REGIONS];
+struct mm_region *mem_map = xen_mem_map;
+
+static int get_next_memory_node(const void *blob, int mem)
+{
+	do {
+		mem = fdt_node_offset_by_prop_value(blob, mem,
+						    "device_type", "memory", 7);
+	} while (!fdtdec_get_is_enabled(blob, mem));
+
+	return mem;
+}
+
+static int setup_mem_map(void)
+{
+	int i, ret, mem, reg = 0;
+	struct fdt_resource res;
+	const void *blob = gd->fdt_blob;
+
+	/*
+	 * Add "magic" region which is used by Xen to provide some essentials
+	 * for the guest: console, xenstore etc.
+	 */
+	xen_mem_map[0].virt = GUEST_MAGIC_BASE;
```

Comment:
I was going to ask how you deal with the cache here, but it looks like _dcache_enable()_ will flush the cache entirely with set/way :(.

I would strongly suggest to avoid set/way as this is a really expensive operation in virtualization (we have to walk the guest memory and flush line by line).

Reply:
This is probably what I see that if given more memory the guests starts slowly